### PR TITLE
cleanup: ppm sc api

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area libscap
/area libsinsp

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

Refactors ppm_sc/tp/events APIs, introducing a new set class by generalizing Falco `falco_event_types` class. 

* properly use ppm_sc and ppme enum (defined as types now) in ppm_sc sinsp API
* added a libsinsp::events:: namespace that holds ppm_sc,tp,events related static functions
* renamed sinsp_{ppm_sc,tp} source files to sinsp_events_{ppm_sc,tp} 
* ported tp enum to a ppm_tp_code type
* added libsinsp::events::set template class

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
refactor(userspace)!: renamed multiple events related API; moreover, added a new libsinsp::events namespace where all the new APIs live. Finally, addeda libsinsp::events::set class to manage the ppm_sc/tp/events types.
```

### CODE REVIEW

**Comment by FedeDP** [comment]
Need to fix libsinsp `interesting_syscalls` and `interesting_tracepoints` tests.

File: userspace/libsinsp/CMakeLists.txt
Comment (jasondellaluce) [inline userspace/libsinsp/CMakeLists.txt:122]: Indentation?

File: userspace/libsinsp/sinsp.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp.h:872]: Moving all these is the biggest breaking change of this PR. Please let's make sure the commit names and release notes have a conventional commit prefix containing `!`, and let's be explicit about the action items for fixing the issues of whoever pulls this upstream!

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:45]: Should we add a constructor for `std::set<ppm_type>` so that we have parity?

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:144]:
```suggestion
	set intersect(const set& other) const
```

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:127]:
```suggestion
	set diff(const set& other) const
```

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:85]: Should we one method for `std::unordered_set<ppm_type>` so that we have parity?

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:145]: I know we have `merge` that does a similar thing, but at this point would it make sense to add `set union(const set& other) const`? That way this class would be all-around versatile.

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [inline userspace/libsinsp/sinsp_events.h:274]: Since we are refactoring this bottom-up, I was thinking that maybe we could make these function names shorter, add the few ones missing, and be more consistent. Here's my proposal, but feel free to ignore it as this is more a matter of taste (the bold ones have not been implemented so far):
- all_sc_set
- all_tp_set
- **all_event_set**
- sinsp_state_sc_set
- sinsp_state_tp_set
- sinsp_state_event_set
- simple_sc_set
- io_sc_set
- io_sc_set
- file_sc_set
- net_sc_set
- proc_sc_set
- sys_sc_set
- sc_set_to_names
- tp_set_to_names
- event_set_to_names
- names_to_sc_set
- **names_to_tp_set**
- **names_to_event_set**
- sc_set_to_event_set
- **event_set_to_sc_set**

**Review by jasondellaluce — COMMENTED** [review]
Love all this! It's now much more clear!

I left you few initial comments, feel free to ignore some of those if you think they're out of context

**Comment by FedeDP** [comment]
@jasondellaluce NOTE: i am still open to update APIs namings, since now we have a proper namespace and proper types.

File: userspace/libsinsp/sinsp_events.h
Comment (incertum) [reply userspace/libsinsp/sinsp_events.h:311]: Agree that `ppm` and `ppme` can be reserved as a concept under the hood, and in the API we use `sc` and `event` only. However, the `enforce` prefix would make it clearer that you can pass a set and add to it, so I would keep that.

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (incertum) [inline userspace/libsinsp/sinsp_events_ppm_sc.cpp:25]: adding to @jasondellaluce point, given we refactor from the ground up, would love to see `scap_get_modifies_state_ppm_sc` be defined in this API and if you agree @FedeDP we might as well do the refactor of only enforcing the EF_MODIFIES_STATE and drop the UF_NEVER_DROP enforcement, but this can maybe be done in a follow up PR to not overload this refactor. Also it's more sinsp modifies state ... so we could rename one more time as well.

**Review by incertum — COMMENTED** [review]
Agree with making all set operations complete to cover all future use cases.

File: userspace/libsinsp/sinsp_events.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h:135]: Yep fully agree!

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events_ppm_sc.cpp:25]: > we might as well do the refactor of only enforcing the EF_MODIFIES_STATE and drop the UF_NEVER_DROP enforcement

Yep, i'd leave this to another PR since it is a somewhat disruptive change!

>  would love to see scap_get_modifies_state_ppm_sc be defined in this API

You mean a `sinsp_modifies_state_set`? This is what `enforce_sinsp_state_ppm_sc` already does, right?

File: userspace/libsinsp/sinsp_events.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h:282]: > However, the enforce prefix would make it clearer that you can pass a set and add to it, so I would keep that.

Given that new `libsinsp::events::set` class provides `merge` method, we could just "enforce" the set for `enforce_simple_ppm_sc_set`, leaving all other sets as just ... sets, without enforcing anything.
Libs client could do something like:
```
io_set = libsinsp::events::io_sc_set();
net_set = libsinsp::events::net_sc_set();

final_set = io_set.merge(net_set);
// use final set
```

Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h:282]: @jasondellaluce love the renamings! Will leave them as latest commit, once all build issues are fixes (macos... :/ )

File: driver/bpf/plumbing_helpers.h
Comment (FedeDP) [reply driver/bpf/plumbing_helpers.h:304]: It would make sense! But we already have:
* ppm_event_category
* ppm_event_entry
* ppm_event_flags
* ppm_event_info

:P

File: userspace/libsinsp/sinsp_events.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h]: Done ;)

File: userspace/libsinsp/sinsp_events.h
Comment (jasondellaluce) [reply userspace/libsinsp/sinsp_events.h:308]: Love the approach of using `merge`! With that, most of these sets may also be made static so that we only need to built them at start time, and reuse them for all the runtime calls.

Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h:308]: Note: i'd avoid adding any new API for now; just rename existing ones; we can later add them as we need.
(eg: `libsinsp::events::all_event_set` would be new)

Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h]: Latest commit updated all function names!

File: userspace/libscap/userspace_flag_helpers.h
Comment (FedeDP) [inline userspace/libscap/userspace_flag_helpers.h:13]: This fixed issues building unit tests on my archlinux laptop.

File: driver/bpf/plumbing_helpers.h
Comment (incertum) [reply driver/bpf/plumbing_helpers.h:304]: I know, we can live with it :)

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (incertum) [reply userspace/libsinsp/sinsp_events_ppm_sc.cpp]: Re your q: `sinsp_state_sc_set()` calls `scap_get_modifies_state_ppm_sc` and not sure why `scap_get_modifies_state_ppm_sc` lives in scap. It appears that we could move all the code directly to `sinsp_state_sc_set()` which may not only simplify everything, but also increase the clarity wrt sinsp <-> scap responsibility separation, something Jason and others brought up too in past threads.

We can chat more about it in the follow up PR :)

**Comment by incertum** [comment]
It's amazing how much more clarity this refactor and renaming achieves, well done 👏 ! Also nice addition `get_sc_set_from_cat`!

@FedeDP may I kindly ask you to consider adding two more mapping functions?
- `names_to_event_set` no current use case, but let's try to be future proof if possible
- `event_set_to_sc_set` adding this as well may give more options to convert the evt types from the filters to state sets

---

Would say let's move forward and get this out of WIP. @Andreagit97 @jasondellaluce @FedeDP what additional testing would you suggest before merging? From my perspective it looks already pretty ready, but let's be extra careful with this critical refactor.

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events_ppm_sc.cpp]: This would make a lot of sense; BUT we first would need to drop `UF_NEVER_DROP` check, since syscall table is not exposed to sinsp ;) 
We can do it later once we decide dropping `UF_NEVER_DROP` set is something we actually want!

**Comment by FedeDP** [comment]
> may I kindly ask you to consider adding two more mapping functions?

Implemented, together with `all_event_set` ;) 

Moreover, updated scap.h related APIs to take an `uint8_t` array (instead of `uint32_t`), and optimized their calls by adding a new `libsinsp::events::set.data()` method that returns underlying `uint8_t` vector data.

File: userspace/libsinsp/sinsp.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp.h:872]: You are right, i will review each commit and add the `!` where needed ;)

File: userspace/libsinsp/CMakeLists.txt
Comment (FedeDP) [reply userspace/libsinsp/CMakeLists.txt]: Done ;)

File: userspace/libsinsp/sinsp.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp.h:872]: Done!

**Comment by FedeDP** [comment]
Latest commit added unit tests for `libsinsp::events::set` class.

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp
Comment (Andreagit97) [inline userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:294]: At the end of our work we should remove all these `ifdef`, we should be able to work on PPM_SC codes without considering the running architecture :tada:

**Review by Andreagit97 — APPROVED** [review]
/approve
Amazing work @FedeDP !!

File: userspace/libsinsp/sinsp_events.h
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.h:300]: Nope, this is the `enforce` one (i decided to leave a `enforce_simple_X_set` under each {tp,sc} categories; i think that a consumer should still be able to use `enforce_simple_sc_set(mySet)` and forget about anything else.

File: userspace/libsinsp/sinsp_events.cpp
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.cpp:94]: Agree!

File: userspace/libsinsp/sinsp_events.h
Comment (Andreagit97) [reply userspace/libsinsp/sinsp_events.h:300]: right now I would keep them, let's say these are our suggested sets to use for libs consumers. Falco won't probably use these APIs anymore.

File: userspace/libsinsp/sinsp_events.cpp
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events.cpp]: I implemented a different logic to avoid adding any new API to sinsp (or continuously calling `scap_get_events_from_ppm_sc` in a loop).

Latest commit implements it + adds a test and + moves all libsinsp::events:: API under libsinsp/events/ folder.

File: userspace/libsinsp/events/sinsp_events.cpp
Comment (FedeDP) [inline userspace/libsinsp/events/sinsp_events.cpp:141]: Directly pass `events_of_interest` underlying vector as const.

File: userspace/libsinsp/events/sinsp_events_ppm_sc.cpp
Comment (FedeDP) [inline userspace/libsinsp/events/sinsp_events_ppm_sc.cpp:244]: Directly pass `ppm_sc_set` underlying vector as const.

**Review by Andreagit97 — APPROVED** [review]
/approve

File: userspace/libsinsp/sinsp_events.h
Comment (incertum) [reply userspace/libsinsp/sinsp_events.h]: Either way fine with me since we now generally adopt an approach of leveraging generic set operations. As we are now breaking everything anyways my preference would be to go all in and just return the simple_set etc in the sinsp API -> that way one consistent behavior and approach across the board.

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (incertum) [reply userspace/libsinsp/sinsp_events_ppm_sc.cpp]: Yup you probably already know my opinion on `UF_NEVER_DROP` :upside_down_face: would urge to get rid of it for the syscalls selection to have proper separation.

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp
Comment (incertum) [inline userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:377]: Add one more test re `names_to_sc_set` and include one non syscall event, e.g. `container` in names as well to test that we only retrieve the correct `sc_set`.

**Review by incertum — COMMENTED** [review]
Focused on scrutinizing the test conditions and left a few comments.

**Review by incertum — APPROVED** [review]
/approve

File: userspace/libsinsp/sinsp_events_ppm_sc.cpp
Comment (FedeDP) [reply userspace/libsinsp/sinsp_events_ppm_sc.cpp]: Well, `UF_NEVER_DROP` is not taken into account unless you enable dropping mode ;) but Falco was just cleaned up about that: https://github.com/falcosecurity/falco/pull/2415

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp
Comment (FedeDP) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:377]: I think we can improve all of these in a subsequent PR ;) we will be working around the new API anyway!

**Pull request merged**
